library/ssl_tls.c

```diff
@@ -7189,4 +7189,137 @@ int mbedtls_ssl_get_handshake_transcript( mbedtls_ssl_context *ssl,
 }
 #endif /* !MBEDTLS_USE_PSA_CRYPTO */
 
+#if defined(MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED)
+/*
+ * Functions for writing supported_groups extension.
+ *
+ * Stucture of supported_groups from RFC8446:
+ *      enum {
+ *          secp256r1(0x0017), secp384r1(0x0018), secp521r1(0x0019),
+ *          x25519(0x001D), x448(0x001E),
+ *          ffdhe2048(0x0100), ffdhe3072(0x0101), ffdhe4096(0x0102),
+ *          ffdhe6144(0x0103), ffdhe8192(0x0104),
+ *          ffdhe_private_use(0x01FC..0x01FF),
+ *          ecdhe_private_use(0xFE00..0xFEFF),
+ *          (0xFFFF)
+ *      } NamedGroup;
+ *      struct {
+ *          NamedGroup named_group_list<2..2^16-1>;
+ *      } NamedGroupList;
+ */
+
+#define SSL_GROUP_IS_UNSUPPORTED    0
+#define SSL_GROUP_IS_ECDHE          1
+#define SSL_GROUP_IS_DHE            2
+static int ssl_check_group_type( const mbedtls_ssl_config *conf,
+                                 const uint16_t group )
+{
+#if defined(MBEDTLS_ECDH_C)
+#if defined(MBEDTLS_SSL_PROTO_TLS1_2)
+    if( mbedtls_ssl_conf_is_tls12_only( conf )
+        && mbedtls_ssl_named_group_is_ecdhe( group ) )
+        return( SSL_GROUP_IS_ECDHE );
+#endif /* MBEDTLS_SSL_PROTO_TLS1_2 */
+
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3)
+    if( mbedtls_ssl_conf_is_tls13_only( conf )
+        && mbedtls_ssl_tls13_named_group_is_ecdhe( group ) )
+        return( SSL_GROUP_IS_ECDHE );
+#endif /* MBEDTLS_SSL_PROTO_TLS1_3 */
+
+#endif /* MBEDTLS_ECDH_C */
+    if( mbedtls_ssl_tls13_named_group_is_dhe( group ) )
+        return( SSL_GROUP_IS_DHE );
+
+    return( SSL_GROUP_IS_UNSUPPORTED );
+}
+
+int mbedtls_ssl_write_supported_groups_ext( mbedtls_ssl_context *ssl,
+                                            unsigned char *buf,
+                                            unsigned char *end,
+                                            size_t *out_len )
+{
+    unsigned char *p = buf ;
+    unsigned char *named_group_list; /* Start of named_group_list */
+    size_t named_group_list_len;     /* Length of named_group_list */
+    const uint16_t *group_list = mbedtls_ssl_get_groups( ssl );
+
+    *out_len = 0;
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3)
+    if( mbedtls_ssl_conf_is_tls13_only( ssl->conf )
```

Comment:
It would be cleaner to not call this function in that case thus to move this check to the TLS 1.3 code calling this function.